crates/ruff_python_formatter/src/expression/expr_joined_str.rs

```diff
@@ -1,21 +1,43 @@
 use crate::context::PyFormatContext;
-use crate::expression::parentheses::{NeedsParentheses, OptionalParentheses};
-use crate::{not_yet_implemented_custom_text, FormatNodeRule, PyFormatter};
-use ruff_formatter::{write, Buffer, FormatResult};
+use crate::expression::parentheses::{NeedsParentheses, OptionalParentheses, Parentheses};
+use crate::expression::string::{StringPrefix, StringQuotes};
+use crate::expression::Expr;
+use crate::prelude::*;
+use crate::{verbatim_text, FormatNodeRule, PyFormatter};
+use ruff_formatter::{write, Buffer, FormatError, FormatResult};
 use ruff_python_ast::node::AnyNodeRef;
 use rustpython_parser::ast::ExprJoinedStr;
 
 #[derive(Default)]
 pub struct FormatExprJoinedStr;
 
 impl FormatNodeRule<ExprJoinedStr> for FormatExprJoinedStr {
-    fn fmt_fields(&self, _item: &ExprJoinedStr, f: &mut PyFormatter) -> FormatResult<()> {
-        write!(
-            f,
-            [not_yet_implemented_custom_text(
-                r#"f"NOT_YET_IMPLEMENTED_ExprJoinedStr""#
-            )]
-        )
+    fn fmt_fields(&self, item: &ExprJoinedStr, f: &mut PyFormatter) -> FormatResult<()> {
+        let ExprJoinedStr { range, values } = item;
+
+        // TODO: can we re-use instead of copy/paste?
```

Comment:
Would it be possible to re-use `FormatStringPart` and add a new field of type `StringKind` that has two variants `Default` and `FString` where the `FString` variant stores the quotes used by the `FString`? 

Re-using `FormatStringPart` has the benefit that it normalizes the string content.

Reply:
maybe i'm misunderstanding your suggestion, but `FormatStringPart` formats "entire" strings. here we only need the string "decorations", (prefixes and quotes)

Reply:
Oh I see. I think copy-pasting is fine. The *hard* part is hidden behind `StringPrefix` and `StringQuotes`. Will it be necessary to determine the preferred quotes as well or do f-string always use the configured quote style?

Reply:
> The hard part is hidden

👍 

> Will it be necessary to determine the preferred quotes as well or do f-string always use the configured quote style?

sorry i don't understand your question

possibly related: where i left this pr last night was with nested quotes outstanding. i haven't yet looked at how this is is handled in regular strings, but i suspect we need something similar

Reply:
oh, interesting. this is about to change with python 3.12 (scheduled for this October) https://peps.python.org/pep-0701/ after which _any valid python expression can be used, so in particular, using the same quotes becomes ok

Reply:
> sorry i don't understand your question

Black prefers using `"` over `'` but not if changing the quotes would require more escape sequences:

```
'abcd' -> "abcd"
'Hy "Micha"' -> remains unchanged, because "Hy \"Micha\""` is worse 
```


The question is. Does this also apply to strings and if so, does it apply to the whole string or only to individual parts?

Reply:
> oh, interesting. this is about to change with python 3.12 (scheduled for this October) https://peps.python.org/pep-0701/ after which _any valid python expression can be used, so in particular, using the same quotes becomes ok

I'd start with the pre 3.12 syntax first (and an assertion that quotes match the older expectations). We can add support for 3.12 in a second step, where we also need to check the python version the user has set.

Reply:
>  start with the pre 3.12 syntax first 

yep makes sense and already the plan

> Does this also apply to strings and if so, does it apply to the whole string or only to individual parts?

do you mean _f -strings?

my understanding so far is that pre 3.12, it works exactly the same because from a quoting perspective the contained FormattedValues don't change things (i.e. the quote rules can be treated as if there are no curly brackets in the f-string. what i'm trying to figure out atm is how to implement this since, unlike regular strings, the inner strings (inside curlies) don't "know" that they are contained in outer strings. i've only just started looking at it, but ideas are welcome

Reply:
> do you mean _f -strings?

yes, I meant f-strings. Grammarly enjoys removing the `f-` from f-string. 

> my understanding so far is that pre 3.12, it works exactly the same because from a quoting perspective the contained FormattedValues don't change things

Does that mean that Black computes the most common quote across all string parts and then normalizes each part?

>  what i'm trying to figure out atm is how to implement this since, unlike regular strings, the inner strings (inside curlies) don't "know" that they are contained in outer strings. i've only just started looking at it, but ideas are welcome

You can implement `FormatRuleWithOptions` on `FormatExprConstant`  and pass down a custom option (layout) telling it it is inside an  f-string.

Reply:
> Does that mean that Black computes the most common quote across all string parts and then normalizes each part

from some more testing:

```
>>>  f"{'\"'}"
SyntaxError: f-string expression part cannot include a backslash
```

maybe this means we can't normalize at all (and don't need to)


> You can implement FormatRuleWithOptions on FormatExprConstant 

ah yes. thanks

Reply:
actually maybe this isn't enough (unless i'm missing something). because the nested string might not be the _immediate child, e.g. in `f"{foo('')}"` .  could we set something in the `PyFormatContext` while inside f-strings?

Reply:
We could, as a last resort, I try to avoid the context when I can because:

* You need to be extremely careful to correctly restore to the previous value (even if an error occurred)
* It complicates error recovery (needs to be snapshotted). This isn't something we do today, but may want to do in the future (requires an error-resilient parser)
* I find it much harder to reason about: Who sets the value? Who reads the value? Is it still used?

Reply:
what's the alternative? add an attribute to all expressions that contain other expressions and make sure it's propagated during formatting? will try that out too if you think that's got legs? also open to other ideas?

Reply:
Yeah, we could pass something like an `ExpressionContext`, that also stores the group id from `NodeLevel::Expression` or whether it is parenthesized. I considered exploring this at some point to see how cumbersome it is/if it improves readability. It wouldn't be necessary to route the context through to all expressions. It is only necessary to pass it through to expressions that contain child expressions or nodes that are e.g. strings that explicitly need the information

I would recommend exploring this in a separate PR

Reply:
sure, will take a look (might come back here and ask for more pointers)

also (in case it ends up related), the "no backslashes inside f-strings" is slightly more subtle than i originally noticed. it's only `ExprFormattedValue`s that can't have backslashes. (i.e. `f" \' {foo}"` is ok) so i think this means the string normalisation now needs to know if there are quotes inside any contained `FormattedValue`s since we can't change the outer quotes if that would require escaping inside a formatted value

Reply:
ok, back to looking at this, could you elaborate on the suggestions of passing an `ExpressionContext` please? not sure i quite understand what you had in mind.

Reply:
Let's go with your initial proposed context-based implementation and see how it turns out. We still have the option to refactor later.

Reply:
oki

related to this, i think i also need to walk all children of some expressions to see if any of them are `FormattedValue`s with quotes inside of them (which we can't escape, so we can't change outside quotes). do we have any helpers to walk expressions and all their children?

Reply:
Not a very good one... You must implement your own `PreorderVisitor`. I would love to have a `children(node)` method that returns an iterator. That method would probably need to allocate a stack internally to queue the next children but it is so much more ergonomic than having to implement a visitor.

Reply:
note that this wasn't needed in the end. immediate children is enough